### ISSUE TITLE
Multimodal C4 not in WebDataset shard format - in jsonl format instead. Any code to convert?

### ISSUE THREAD

**Issue by roboswell** [mlfoundations/open_flamingo#168]
I noticed that all datasets for Open Flamingo need to be in WebDataset shard format. On your GitHub page, you mention that the Multimodal C4 dataset should automatically be in WebDataset shard format. But on the mmc4 GitHub page, when I downloaded it, it downloads in purely jsonl format.

Is there code you could recommend for properly converting mmc4 jsonl files into WebDataset shards?

**Comment by anas-awadalla** [comment]
Yes sorry forgot to update Readme. Yes working on cleaning up script and will share in a little bit.

**Comment by roboswell** [comment]
Thanks, I really appreciate it!

**Comment by XinyangHan** [comment]
Looking forward to that script!!

**Comment by anas-awadalla** [comment]
The script can be found [here](https://github.com/mlfoundations/open_flamingo/blob/main/open_flamingo/train/convert_mmc4_to_wds.py)

**Comment by roboswell** [comment]
Hi Anas,

As I've been trying to run your script, I've encountered a potential
problem. On Allen AI's mmc4 GitHub repo, while it does provide a url for
downloading zipped jsonl files for the dataset, your script seems to
require tar-ed images corresponding to the jsonl files as well. However,
the mmc4 GitHub repo does not provide any url for downloading images in tar
files.

Can you provide the url I would need to go to download the tar files?

Thank you,
Rob Boswell

**Comment by anas-awadalla** [comment]
@roboswell Ok yes here I am assuming you have access to the raw images. It seems that using the clip features will be more useful for folks. I need to change up our training code to support directly using clip features.

**Comment by roboswell** [comment]
I don't know that I would be using the CLIP features - except in the
context of the pretrained Open Flamingo Vision Encoder. When you refer to
CLIP features, are you referring to something else besides the Vision
Encoder?

Would it be possible to create a script that simply extracts the raw images
from the raw image urls ("raw_url") provided in the jsonl files once
downloaded?

**Comment by anas-awadalla** [comment]
Yes clip features would replace forward pass through OpenFlamingo VE. Yeah I can also use the image urls. It will be slow but perhaps more useful outside OpenFlamingo.

**Comment by roboswell** [comment]
I'm just brainstorming, but if in fact there is a way to allow users to
download the raw images (if you currently have them available to be placed
into tar files), then perhaps it would make more sense in the script to
recursively download them directly from the server rather than having users
extract them from the "raw_url" field in the jsonl files once they are on
their local machines, due to the time factor.

**Comment by roboswell** [comment]
Hi @anas-awadalla, I was wondering whether progress has been made on the convert_mmc4_to_wds.py script for Open Flamingo? I'm also trying to better understand what the script will be doing once complete. Will it 1) output document shards from the jsonl files? Will it also 2) take raw images and create image shards compressed in tar files? And 3) will it output CLIP features in order to **pretrain** the Vision Encoder? 

Based on your prior response to my post, I am a bit confused why the script would seek to incorporate CLIP features in order to convert MMC4 to WedDataset format for the following reasons: in Deep Mind's Flamingo paper, they do not use the M3W dataset (which MMC4 is meant to replicate) during pre-training - only training. On the other hand, CLIP features (or more specifically contrastive loss trained features) from two image-text paired datasets (the equivalent of which would be LAION-2B for Open Flamingo if I understand correctly) were used for pre-training the Vision Encoder. So, if MMC4 is only meant for training and if LAION-2B is meant for both pretraining and training, why then are you integrating CLIP features into the script for MMC4? Am I misunderstanding something?

**Comment by anas-awadalla** [comment]
Hi @roboswell, sorry for the delay! I will wrap this up by Thursday. I think my previous response was not clear so sorry for the confusion. Images for MMC4 are provides as CLIP features so if you want access to the raw images you will need to download them. There is a [script](https://github.com/allenai/mmc4/issues/7#issuecomment-1522275462) in the MMC4 repo for doing so! I am not sure I understand the distinction between pre-training and training stages for Flamingo. Do you mean pre-training a vision encoder from scratch? Can you elaborate on that? Since the CLIP encoder is frozen for Flamingo models, the CLIP features from MMC4 can be used for training as they just replace the forward pass through the CLIP encoder.

To answer your other questions:
> Will it 1) output document shards from the jsonl files?
Yep! We will output webdatatset shards.
> Will it also 2) take raw images and create image shards compressed in tar files? 
No, it will instead contain the CLIP features.
> And 3) will it output CLIP features in order to pretrain the Vision Encoder?
While you will have the CLIP features I don't think you can use those to do contrastive pre-training. Let me know if I misunderstood this question.

**Comment by roboswell** [comment]
Hi @anas-awadalla, thanks for clarifying! I think my confusion comes from my understanding (whether correct or incorrect) that M3W basically consists of a ton of interleaved web pages with the raw images on each web page just being replaced by a tag. So, I imagined that if MMC4 was just like M3W then during the training phase there would be actual raw images being fed through the Vision Encoder at the same time as text corresponding to each image were tokenized and then fed through the LLM, and that only after the raw images were fed through the Vision Encoder would there be CLIP features for the images that would then be passed on to the Perceiver Resampler.

My understanding from the Flamingo paper is that the difference between pretraining and training is that during pretraining - that is where the actual raw images from DeepMind's equivalent of the LAION-2B dataset are fed through the Vision Encoder while its corresponding texts are processed through a BERT encoder, and consequently that is the only time contrastive loss is minimized/learning occurs for the Vision Encoder, since the Vision Encoder is - like you said - frozen during the training process. But, it was also my understanding that - during the training phase - that is when you would place raw images from both the DeepMind equivalent datasets of LAION-2B as well as the M3W (MMC4 in your case) interleaved dataset through the Vision Encoder and that you would end up with CLIP features being extracted from the raw image pixels as a result of training since that is the purpose of the Vision Encoder-to take pixels and extract CLIP features.

So, my confusion comes from the idea that with MMC4 the features have already been extracted from the Vision Encoder even before the "training phase" has occurred. So, it sounds like it is just that you have already trained the MMC4 dataset on the complete Open Flamingo 9B model (Vision Encoder, Perceiver Resampler, Gated Cross Attention blocks, and LLM), and so now there is no need to go through the same process all over again for when you train the higher parameter versions of Open Flamingo of feeding the original raw images through the Vision Encoder and re-extracting the CLIP features? I.e., you're making life easier for anyone seeking to use MMC4 for future training of models.

As an aside, I am curious about the image resolution that was used to pretrain the CLIP Vision Encoder you used for your 9B model (ViT L/14, I think). Did OpenAI use 336 pixels for pretraining ViT L/14? And for your training phase of the Open Flamingo 9B model, did you use 224 pixels? The reason I ask is that in DeepMind's Flamingo paper, I noticed they pretrained their Vision Encoder with 288 pixels, and when they trained the entire model with the Vision Encoder frozen they used a **_higher_** resolution of 320 pixels. And I just wonder if the difference between the pretraining resolution vs. the training resolution could impact the model's performance? E.g., if correct that the OpenAI ViT L/14 model was pretrained at 336 pixels and you lower the resolution to 224 pixels for training Open Flamingo, could the performance of Open Flamingo end up being very different than DeepMind's Flamingo model since they actually increased their resolution for training compared to when they pretrained their Vision Encoder?

**Comment by anas-awadalla** [comment]
> Correct me if I'm wrong, but it sounds like the reason why you're using CLIP features rather than the raw images is because you have already processed MMC4's original raw images through the Vision Encoder while training the Open Flamingo 9B model, and since you only need to process the raw images once through the Vision Encoder to obtain the CLIP features, it just saves people time who want to use MMC4 for training models to use the CLIP features that have already been extracted rather than re-processing the raw images through the Vision Encoder each time they want to train a new model.

Exactly! Yes this sounds right.

> My understanding of the difference between the Flamingo paper's pretraining vs. training phase is that during pretraining, the authors took their image-text pair datasets (equivalent of LAION-2B) and used contrastive training with their Vision Encoder and a BERT encoder, then discarded the BERT encoder after pretraining. After pretraining was completed is when the Vision Encoder's weights and biases were frozen, and so - like you said - there is no parameter updates of the Vision Encoder during the regular "training phase" of the overall Flamingo model.

We decided not to train our own vision encoder. There are already many great ones on OpenCLIP we can just plug in. So I think this is where I wasn't clear about things. Since we use an off the shelf CLIP encoder we can just skip the 'pre-training' stage mentioned in the paper.

> Actually, I also have a couple of additional questions. I noticed that your training resolution for the raw images appears to have been 224 x 224. Is that right? And, if I understand correctly, the ViT L/14 image encoder was originally pretrained by OpenAI on (I think) 336 x 336 pixels. If that's the case, then your resolution decreased between pretraining and training. But, I also noticed in the Flamingo paper that the authors pretrained their Vision Encoder with 288 x 288 pixels, and then trained their whole model (with Vision Encoder and LLM frozen) with a higher resolution of 320 x 320 pixels. On pg. 21 the authors write, "The increase in resolution during the final stage training was motivated by Touvron et al. (2019) that show one can obtain improve performance at a higher test-time resolution when using CNNs." I'm wondering if you have looked at whether the improved performance effect of increasing the resolution at training time still holds with Vision Transformers as the Vision Encoder?

Thanks for catching this. I did not know that we might have decreased the resolution and will take another look at this. You are right that they did increase the resolution during training and this is something on my todo list. I will let you know what I find.

> One other thing I've noticed - I can't figure it out - is that even when I attach CUDA to your Open Flamingo model initialization and inference code at the appropriate places, the code won't run on GPUs, and so it ends up taking about an hour and a half just to load the checkpoint.pt to the model via Pytorch's load_state_dict() function. Do you know why this might be happening, whether given the Open Flamingo source code I may need to do something more than attaching .to("cuda") to connect with GPUs?

Hmm that is weird. There is a known [issue](https://github.com/mlfoundations/open_flamingo/issues/160) where the tokenizer takes around a minute to load but this seems different. Can you open an issue with details about your setup so we can see if others are facing the same issue? I am also wondering once the model is loaded is it excessively slow?

**Comment by roboswell** [comment]
@anas-awadalla Yeah, it is a strange error. I think it's possible that it might be something wrong with the docker container I'm running it in, so I'll try to investigate further. I'm pretty sure this problem I've been encountering is not the same as the tokenizer issue, but I'll try to verify that as well.

Do you know when the CLIP features will be available to download so that the script that converts the jsonl files to WebDataset format (https://github.com/mlfoundations/open_flamingo/blob/main/open_flamingo/train/convert_mmc4_to_wds.py) will be able to be run? It looks like right now, the script still requires the "--image_shards" argument (which from your earlier explanation sounds like its referring to the raw images). Once the CLIP features are able to be downloaded, will the convert_mmc4_to_wds.py script be edited for people to be able to include the CLIP features instead of the raw image shards?

**Comment by anas-awadalla** [comment]
As mentioned [here](https://github.com/allenai/mmc4/issues/5#issuecomment-1553560868) the CLIP features we provide are not sufficient for flamingo training. We suggest downloading the raw images for mmc4.

**Comment by anas-awadalla** [comment]
As for the other issues I don’t think the slow downloading is a problem on our end. This line just downloads the weights from the huggingface hub. Can you open a separate issue for for your gpu problem?

**Comment by roboswell** [comment]
> As mentioned [here](https://github.com/allenai/mmc4/issues/5#issuecomment-1553560868) the CLIP features we provide are not sufficient for flamingo training. We suggest downloading the raw images for mmc4.

@anas-awadalla 
I have downloaded raw images using your recommended script ([here](https://github.com/allenai/mmc4/blob/main/scripts/download_images.py)). However, the script to convert mmc4 to WebDataset format ([here](https://github.com/mlfoundations/open_flamingo/blob/main/open_flamingo/train/convert_mmc4_to_wds.py)) requires tar files - not image files. 

In addition, the contents of the convert_mmc4_to_wds.py confuses me because the argument "--image_shards" requires {0..23098} tar files while the argument "--doc_shards" also requires {0..23098} jsonl.zip files. Does this mean there will literally be 1 image per jzonl file? Since each single jsonl contains many web page documents within it, shouldn't there also be many image files per jsonl file?

Are either or both of the above scripts incorrect as presently constituted?

Thank you

**Comment by jerwelborn** [comment]
any reason `convert_mmc4_to_wds.py` was removed or is there still a deep link that can be shared?
